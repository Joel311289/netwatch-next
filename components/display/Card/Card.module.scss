/** Card (main component) **/

.wrapper {
  overflow-y: clip;
  display: block;
  position: relative;
  transition: width 0.2s ease-in, height 0.2s ease-in, border 0s;
  width: fit-content;

  &:not(:empty) {
    border: 1px solid rgb(var(--rgb-color-border-portal) / 90%);
  }

  &:hover {
    box-shadow: var(--box-shadow-card);
  }

  & > a {
    width: auto;
  }
}

/** CardImage (child component) **/

.image {
  transition: all 0.2s ease-in;

  &.linkable {
    cursor: pointer;
  }

  &:not(:empty):hover,
  &.linkable:hover {
    &::before {
      @include box(100%, 100%);

      content: '';
      display: block;
      z-index: 5;
      background-image: linear-gradient(0deg, rgb(20 20 20 / 85%), transparent 200px);
    }
  }
}

/** CardAction (child component) **/

.action {
  box-sizing: border-box;
  padding: 0 !important;
  height: fit-content;
  border-radius: 50%;

  &:hover {
    background-color: rgb(var(--rgb-color-white) / 40%) !important;
  }
}

.actionIcon {
  @include font($size: 30px, $color: var(--color-button-text));

  font-size: 2.5rem !important;
  border: 2px solid var(--color-button-text) !important;
  border-radius: 50%;
  padding: 5px;
}

/** CardActions (child component) **/

.actions {
  @include absolute(bottom 5px right 0);

  background-color: transparent;
  z-index: 2;
  display: none !important;
  padding: 12px;

  .image:hover & {
    display: flex !important;
  }
}

/** CardBody (child component) **/

.body {
  @include absolute(top 0);

  padding: 10px;
}

.heading {
  @include max-lines(2);
  @include box(fit-content, auto);

  color: var(--color-button-text);
  background-color: rgb(0 0 0 / 50%) !important;
  transition: all 0.2s ease-in;
  font-weight: 500;
}
