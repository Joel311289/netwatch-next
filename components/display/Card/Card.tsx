/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import { createContext, useContext, MouseEvent, MouseEventHandler } from 'react';
import { IconType } from 'react-icons';
import { ElementChildren, ElementHTML, ElementLink, ElementSkeleton } from '@/types';
import { withChildrenFiltered } from '@/hoc/withChildrenFiltered';
import { Space } from '@/components/layout';
import { Image } from '@/components/media';
import { Button } from '@/components/forms';
import { Heading } from '@/components/typography';
import { ImageProps } from '@/components/media/Image/Image';
import { classes } from '@/utils/helpers';
import styles from '@/components/display/Card/Card.module.scss';

/* -------------------------------------------------------------------------- */
/** Context **/
/* -------------------------------------------------------------------------- */

type CardContextProps = typeof defaultValue & ElementSkeleton & ElementLink;

const defaultValue = {};

const CardContext = createContext<CardContextProps>(defaultValue);

const useCardContext = () => useContext(CardContext);

/* -------------------------------------------------------------------------- */
/** CardImage (child component) **/
/* -------------------------------------------------------------------------- */

export type CardImageProps = ImageProps;

const CardImage = (props: CardImageProps) => {
  const { href, skeleton } = useCardContext();

  return (
    <Image
      alt="image"
      skeleton={skeleton}
      href={href}
      {...props}
      className={classes(styles.image, href && styles.linkable, props.className)}
    />
  );
};

/* -------------------------------------------------------------------------- */
/** CarAction (child component) **/
/* -------------------------------------------------------------------------- */

type CardActionProps = ElementHTML & {
  icon: IconType;
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const CardAction = ({ className, icon: Icon, onClick }: CardActionProps) => {
  useCardContext();

  const handleClick = (event: MouseEvent<HTMLButtonElement>): void => {
    onClick && onClick(event);
  };

  return (
    <Button
      clear
      ariaLabel="action"
      className={classes(styles.action, className)}
      onClick={handleClick}>
      <Icon className={styles.actionIcon} />
    </Button>
  );
};

/* -------------------------------------------------------------------------- */
/** CarActions (child component) **/
/* -------------------------------------------------------------------------- */

type CardActionsProps = ElementHTML & ElementChildren<JSX.Element | Array<JSX.Element>>;

const CardActions = ({ className, children }: CardActionsProps) => {
  useCardContext();

  return (
    <Space justify="end" gap={8} className={classes(styles.actions, className)}>
      {children}
    </Space>
  );
};

const CardActionsWithChildrenFiltered = withChildrenFiltered(CardActions, {
  Item: CardAction
});

/* -------------------------------------------------------------------------- */
/** CardBody (child component) **/
/* -------------------------------------------------------------------------- */

type CardBodyProps = ElementHTML & {
  title: string;
  description?: string;
  maxLines?: number;
};

const CardBody = ({ className, title }: CardBodyProps) => {
  const { href, skeleton } = useCardContext();

  if (skeleton || !title) {
    return null;
  }

  return (
    <Space direction="column" gap={2} className={classes(styles.body, className)}>
      <Heading href={href} className={classes(styles.heading, href && styles.linkable)}>
        {title}
      </Heading>
    </Space>
  );
};

/* -------------------------------------------------------------------------- */
/** Card (main component) **/
/* -------------------------------------------------------------------------- */

export type CardProps = typeof defaultProps &
  ElementHTML &
  ElementChildren<Array<JSX.Element>> &
  ElementSkeleton &
  ElementLink & {
    onClick: MouseEventHandler<HTMLDivElement>;
  };

const defaultProps = {
  children: []
};

const Card = ({ children, className, skeleton, href, onClick }: CardProps) => {
  return (
    <CardContext.Provider value={{ href, skeleton }}>
      <div className={classes(styles.wrapper, className)} onClick={onClick}>
        {children}
      </div>
    </CardContext.Provider>
  );
};

Card.defaultProps = defaultProps;

const CardWithChildrenFiltered = withChildrenFiltered(Card, {
  Image: CardImage,
  Body: CardBody,
  Actions: CardActionsWithChildrenFiltered
});

export default CardWithChildrenFiltered;
