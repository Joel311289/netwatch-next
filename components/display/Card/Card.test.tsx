import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IoMdAdd } from 'react-icons/io';
import Card from '@/components/display/Card/Card';

describe('Tests Card component', () => {
  test('renders component correctly', () => {
    const { container } = render(
      <Card>
        <Card.Image src="https://picsum.photos/200/300" width={50} ratio={1} />
        <Card.Body title="Title" description="Description" />
      </Card>
    );

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    render(
      <Card>
        <Card.Image src="https://picsum.photos/200/300" width={50} ratio={1} />
        <Card.Body title="Title" />
      </Card>
    );

    expect(screen.getByLabelText('image')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
  });

  test('works click on action image ', async () => {
    const handleClick = jest.fn();

    render(
      <Card>
        <Card.Image src="https://picsum.photos/200/300" width={50} ratio={1}>
          <Card.Actions>
            <Card.Actions.Item onClick={handleClick} icon={IoMdAdd} />
          </Card.Actions>
        </Card.Image>
      </Card>
    );

    userEvent.click(screen.getByLabelText('action'));

    await waitFor(() => {
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });
});
