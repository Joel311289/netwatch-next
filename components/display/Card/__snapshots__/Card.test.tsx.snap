// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Card component renders component correctly 1`] = `
<div>
  <div
    class="wrapper wrapper"
  >
    <div
      class="wrapper"
      role="img"
    >
      <div
        aria-label="image"
        class="image image"
      />
      <style>
        
        .wrapper {
          width: 50px;
          height: 50px;
          min-width: 50px;
        }
        .wrapper::before {
          padding-top: 100%;
        }

        .image {
          background-image: url(https://picsum.photos/200/300);
        }
      
      </style>
    </div>
    <div
      class="wrapper body"
    >
      <h2
        class="wrapper heading"
      >
        Title
      </h2>
      <span
        class="wrapper description sm"
      >
        Description
        <style>
          
        wrapper {
          
        }
      
        </style>
      </span>
      <style>
        
        .wrapper {
          display: flex;
          flex-direction: column;
          flex-wrap: nowrap;
          gap: 3px;
          
          
        }
      
      </style>
    </div>
  </div>
</div>
`;
